core/services/balance_monitor.go

```diff
@@ -0,0 +1,129 @@
+package services
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/chainlink/core/assets"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/eth"
+	"github.com/smartcontractkit/chainlink/core/store"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+
+	gethCommon "github.com/ethereum/go-ethereum/common"
+	"github.com/pkg/errors"
+)
+
+// BalanceMonitor checks the balance for each key on every new head
+type BalanceMonitor interface {
+	store.HeadTrackable
+}
+
+type balanceMonitor struct {
+	store             *store.Store
+	gethClientWrapper store.GethClientWrapper
+	ethBalances       map[gethCommon.Address]*assets.Eth
+	ethBalancesMtx    *sync.Mutex
+}
+
+// NewBalanceMonitor returns a new balanceMonitor
+func NewBalanceMonitor(store *store.Store, gethClientWrapper store.GethClientWrapper) BalanceMonitor {
+	return &balanceMonitor{
+		store:             store,
+		gethClientWrapper: gethClientWrapper,
+		ethBalances:       make(map[gethCommon.Address]*assets.Eth),
+		ethBalancesMtx:    new(sync.Mutex),
+	}
+}
+
+// Connect complies with HeadTrackable
+func (bm *balanceMonitor) Connect(_ *models.Head) error {
+	// Connect head can be out of date, so always query the latest balance
+	bm.checkBalance(nil)
+	return nil
+}
+
+// Disconnect complies with HeadTrackable
+func (bm *balanceMonitor) Disconnect() {}
```

Comment:
What do you think if Disconnect terminates all the goroutines started by checkBalance()?

Reply:
It _could_.

To be honest, `Disconnect` is kind of a hack anyway. We should really be using contexts everywhere.

Remember we spoke about this before? Here is the pivotal tracker story https://www.pivotaltracker.com/n/projects/2441917